### ISSUE TITLE
FNA++ ResNet Search Space and group remapping

### ISSUE THREAD

**Issue by Hrayo712** [JaminFong/FNA#20]
Hi @JaminFong 

I hope you're doing well. I am trying to reproduce the ResNet-based Search space from the FNA++ paper. I have a couple of questions. On the paper (Table 7) you do not include the Skip connection. Is this still part of the search space for the normal primitives? 

Also, how many layers do you allow per block ?
e.g., SSDLite/RetinaNet on the Mobilenetv2-based search space was 
`num_layers= [4, 4, 4, 4, 4, 1]`

**Comment by Hrayo712** [comment]
Hey, @JaminFong 

I am also a bit confused on how the group remapping should be implemented. Your remapping on width and kernel level is implemented as follows:

`mapped_dict[k].narrow(0, 0, output_dim).narrow(1, 0, input_dim).narrow(
                    2, (w_model - w_min) // 2, w_min).narrow(3, (h_model - h_min) // 2, h_min).copy_(
                    v.narrow(0, 0, output_dim).narrow(1, 0, input_dim).narrow(
                    2, (w_pre - w_min) // 2, w_min).narrow(3, (h_pre - h_min) // 2, h_min))`

I modified adding the following:
`nr_groups =  v.shape[1] // model_dict[k].shape[1]`
  `in_dim_slice = input_dim // nr_groups`

`mapped_dict[k].narrow(0, 0, output_dim).narrow(1, 0, in_dim_slice).narrow(
                    2, (w_model - w_min) // 2, w_min).narrow(3, (h_model - h_min) // 2, h_min).copy_(
                    v.narrow(0, 0, output_dim).narrow(1, 0, in_dim_slice).narrow(
                    2, (w_pre - w_min) // 2, w_min).narrow(3, (h_pre - h_min) // 2, h_min))`

So for example: if the original convolution is (512, 512, x, x) and in the grouped convolution g = 2 (512, 256, x, x) Then I would be copying only the first 256 from the original one, onto the new one in the supernetwork. Is this how it is supposed to be ?

**Comment by JaminFong** [comment]
Hi, @Hrayo712 ,
1. The skip connection is not included in the ResNet-based search space.
2. The maximum number of layers in the search space is set the same as the original ResNet networks, _i.e._, `[3, 4, 6, 3]` for Res-50 and `[2, 2, 2, 2]` for Res-18.
3. Please refer to Eq. 4 in FNA++ for the width-level parameter remapping for grouped convolutions.
Hoping my answer can help you.

Best.

**Comment by Hrayo712** [comment]
Hi @JaminFong ,

Thanks for the reply! 

I believe I have followed equation 4 correctly from your paper. However, after the search, I observe that the additional blocks in the search space are only used on the last block of the architecture, all the other blocks are populated with `k3_g1`. Have you observed this kind of behavior before ?

I wonder if you could share the encoding of the architecture which gave you the performance you report on the paper. Also, if you could share the remapping code for ResNet would be great, even if it isnt clean. It could help me identify if I might be doing something wrong.

I really appreciate the support!

Thanks!

**Comment by JaminFong** [comment]
Hi @Hrayo712 ,

> the additional blocks in the search space are only used on the last block of the architecture, all the other blocks are populated with k3_g1

I encounter this phenomenon when searching on ResNet-18, but not on ResNet-50. I think this may be due to the search space of Res-18 is too small.
Here is a code segment for width-level remapping on grouped convolutions. You could refer to it if needed.
```
v_s = v.shape
vn_s = v_new.shape
groups = int(v_s[1] / vn_s[1])

v_mapped = v[:vn_s[0]//groups, :vn_s[1], :, :]
for g in range(1, groups):
    v_mapped = torch.cat(
        (v_mapped, v[g*vn_s[0]//groups:(g+1)*vn_s[0]//groups, g*vn_s[1]:(g+1)*vn_s[1], :, :]), 0)
```
`v` denotes the seed parameter, and `v_new` denotes the new parameter to be mapped to.

**Comment by Hrayo712** [comment]
Hi @JaminFong !,

Thanks a lot! this really does help :)

**Comment by Hrayo712** [comment]
Hi @JaminFong ,

Thanks again for all your support. Could you share with me the encoding of the FNA REsNet-50 architecture that you report on the paper, which gave you 36.8 mAP ?
